contracts/DispatcherManager/DispatcherManager.sol

```diff
@@ -0,0 +1,120 @@
+//SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.9;
+
+import {DataTypes} from "../WithSig/DataTypes.sol";
+import {WithSigEnabled} from "../WithSig/WithSigEnabled.sol";
+import {IAddressManager} from "../Config/IAddressManager.sol";
+import {DispatcherManagerStorageV1} from "./DispatcherManagerStorage.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {ReentrancyGuardUpgradeable} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+contract DispatcherManager is
+    WithSigEnabled,
+    Initializable,
+    ReentrancyGuardUpgradeable,
+    DispatcherManagerStorageV1
+{
+    /// @dev Event emitted when a dispatcher is added
+    event DispatcherAdded(address account, address addedDispatcher);
+
+    /// @dev Event emitted when a dispatcher is removed
+    event DispatcherRemoved(address account, address removedDispatcher);
+
+    /// @dev initializer to call after deployment, can only be called once
+    function initialize(IAddressManager _addressManager) public initializer {
+        require(address(_addressManager) != address(0x0), ZERO_INPUT);
+        __ReentrancyGuard_init();
+        addressManager = _addressManager;
+    }
+
+    function _addDispatcher(address _account, address _dispatcher) internal {
+        require(
+            dispatchersByAccount[_account][_dispatcher] == false,
+            DISPATCHER_ALREADY_ASSIGNED
+        );
+        dispatchersByAccount[_account][_dispatcher] = true;
+        emit DispatcherAdded(_account, _dispatcher);
+    }
+
+    function _removeDispatcher(address _account, address _dispatcher) internal {
+        require(
+            dispatchersByAccount[_account][_dispatcher] == true,
+            DISPATCHER_ALREADY_UNASSIGNED
+        );
+        dispatchersByAccount[_account][_dispatcher] = false;
+        emit DispatcherRemoved(_account, _dispatcher);
+    }
+
+    // !!! WARNING !!!
+    // ONLY CALL THIS FUNCTION INSIDE A CONTRACT METHOD
+    // THAT HAS BEEN GUARDED AGAINST REEENTRY
+    // ie, require(msg.sender == tx.origin, "ReentrancyGuard: reentrant call")
+    function callerIsAccountHolderOrDispatcher(address _account)
+        external
+        view
+        returns (bool)
+    {
+        return (_account == tx.origin || // is account holder
```

Comment:
Curious why this is using tx.origin?  Does this mean that a multisig contract that holds NFTs can't use the dispatcher setup?  I guess this makes sense (since a multisig can't sign something) but want to make sure this is expected.

Reply:
The calling context for `DispatchManager#callerIsAccountHolderOrDispatcher()` in a method like `MakerRegistrar#registerNftAsDispatcher()` was down the following chain...

Dispatcher (EOA) `tx.origin` -> MakerRegistrar `msg.sender` -> DispatcherManager

I've removed this method and opted to do a check on the dispatchers mapping directly within the functions that require it.

This will allow for multisig dispatchers as you've suggested.